Build NuGet.Client even if response file is detected

Nirmal4G opened the pull request:
## Bug

Fixes: NuGet/Home#9973
Fixes: NuGet/Home#10139
Regression: No

## Fix

Details:

 - Checkout bash script files with Unix line ending (only LF) even on Windows. This makes sure that we can use the repo in WSL also.

 - MSBuild picks up response files automatically, which messes up the version logic reading from standard output. Adding `-noAutoRsp` to it is one solution but you have to add it to every MSBuild call to get 
   uniform behavior. So, I have refactored the logic to work without needing it.

 - Also, `build.sh` currently downloads dotnet cli v1.0 for bootstrapping and runs MSBuild to get the dotnet cli version to build and test against. Since, v1 is out of support, the Sdk is not available to download. We 
   have to update the version to 2.2, similar to what is specified in `runFuncTests.sh`. Because of the nature of the change in the POSIX scripts, we can no longer depend on `space` character for splitting the branch 
   and version string in the `CliBranchForTesting` property in `build\config.props`. Instead, I have used colon `:` as a separator.

   For Example:
   ```xml
   <CliBranchForTesting Condition="'$(CliBranchForTesting)' == ''">release/5.0.2xx:5.0.202</CliBranchForTesting>
   ```

**NOTE:** This patch was separated from #3270.

## Testing/Validation

Tests Added: No
Reason for not adding tests: Build Infra changes only
Validation: Through successful CI

Nirmal4G commented:
This also includes changes from PR #3552. I included that just to validate my changes via CI. I'll rebase this, once @heng-liu's patch is merged.

zivkan commented:
Given this PR depends on #3552, I'm marking this PR as draft to signal that it's not as high priority to review just yet. I spoke with @heng-liu about that PR, and it's a little bit problematic because of net5 breaking changes and CI agents not having preview VS, but we think we have a path forward before VS 16.8 goes GA.

Please mark the PR as ready to review once #3552 is merged and you rebase your changes.

zivkan commented on build/common.ps1:
```diff
@@ -157,95 +157,81 @@ Function Install-DotnetCLI {
-        else {
-            $Version = $CliChannelAndVersion[1].trim()
-        }
+    $Arch = if ([Environment]::Is64BitOperatingSystem) { "x64"; } else { "x86"; }
```
I think the only place this is used is when calling dotnet-install.ps1, and that script can already figure out the right architecture, so I think we can get rid of this.

zivkan commented on build/config.props:
```diff
@@ -33,12 +33,13 @@
+    <OverrideCliBranchForTesting Condition="'$(LockSDKVersion)' == 'true'">master:5.0.100-preview.8.20417.9;3.1</OverrideCliBranchForTesting>
+    <OverrideCliVersionForBuilding Condition="'$(LockSDKVersion)' == 'true'">"master 5.0.100-preview.8.20417.9"</OverrideCliVersionForBuilding>
+    <CliVersionForBuilding Condition="'$(OverrideCliVersionForBuilding)' != ''">$(OverrideCliVersionForBuilding)</CliVersionForBuilding>
+    <CliVersionForBuilding Condition="'$(OverrideCliVersionForBuilding)' == ''">5.0;3.1</CliVersionForBuilding>
```
NuGet 5.8 is no longer inserting into .NET Core 3.1, so shouldn't test against it, or need it for building. https://github.com/NuGet/NuGet.Client/pull/3570

Also, we can't build with the latest .NET 5 preview 8 or later, as it needs msbuild 16.8 or later, which is only available with preview VS 16.8, which is not available on our CI, so we're limited to .NET 5 SDK preview 7 at max, until VS 16.8 goes GA and is available on our CI agents 😞   But I think this is what you're getting from Heng's PR, so once hers is merged, I don't think you'll need these changes in your branch.

zivkan commented:
@Nirmal4G PR #3552 was closed, because of CI infrastructure issues. Basically, we can't implement those changes until VS 16.8 goes GA, and our CI machines are updated to that version. So, we should be able to in a few weeks, but not yet.

If you can modify your branch so that it does not depend on the changes in #3552, then we can consider taking your changes. Thanks.

Nirmal4G committed:
Always checkout bash files with LF line endings

Adding "eol" to git attributes file for BASH script files
to always checkout with LF line endings

Nirmal4G committed:
Add error checking to bash scripts

Fail early if the external commands encounter an error

Nirmal4G committed:
Fix .NET install when an RSP file is encountered

MSBuild picks up response files automatically,
which messes up reading from standard output.

This fixes the way it's read from std output,
so that it always gets the intended value.

Also, this is just a simplified hack and in no-way
it can get the right value from any MSBuild target
output stream.

Nirmal4G committed:
Match the logic from functional tests script

The current logic assumes there will be no multi-value
dotnet testing branches but we could be wrong...

So, to be on the safe side, It's better to match the logic
with functional tests' version.

Nirmal4G commented:
@zivkan Can you kick off the CI? local build was successful. Let's see if CI reports the same...

Nirmal4G commented:
I don't know what went wrong, here are my test results: [TestResults.zip](https://github.com/NuGet/NuGet.Client/files/5323382/TestResults.zip)

zivkan commented on build/config.props:
```diff
@@ -33,12 +33,13 @@
+    <OverrideCliBranchForTesting Condition="'$(LockSDKVersion)' == 'true'">master:5.0.100-rc.1.20452.10</OverrideCliBranchForTesting>
+    <OverrideCliVersionForBuilding Condition="'$(LockSDKVersion)' == 'true'">5.0.100-rc.1.20452.10</OverrideCliVersionForBuilding>
+    <CliVersionForBuilding Condition="'$(OverrideCliVersionForBuilding)' != ''">$(OverrideCliVersionForBuilding)</CliVersionForBuilding>
+    <CliVersionForBuilding Condition="'$(OverrideCliVersionForBuilding)' == ''">5.0</CliVersionForBuilding>
```
This isn't going to work. This will install the latest 5.0 SDK, which is obviously a higher version that preview 8. Preview 8 onwards needs MSBuild 16.8, but our CI agents only have 16.7. So, even if we fix the other build errors, our Build_and_UnitTest jobs will fail with a error message something like "The .NET SDK requires MSBuild 16.8 or higher".

Nirmal4G replied:
I just found that, fixing now! Sorry about that. My bad!!

Nirmal4G replied:
> "The .NET SDK requires MSBuild 16.8 or higher"

The Core MSBuild 16.8 is bundled with .NET SDK itself, right? I got my local build succeed without it being present in VS (the NETFX version) install.

zivkan replied:
Some of our VS components can only be built with VS's msbuild.exe, and they have ProjectReferences to the projects targeting net5.0. Hence we need msbuild.exe to be able to build the net5.0 (or netcoreapp5.0) projects. Once VS 16.8 goes GA and our CI agents get updated, then we can switch everything over, and it'll be much easier. Until then, we have ugly hacks.

An alternative is to build the `src\NuGet.Core` projects with the .NET SDK and `src\NuGet.Clients\`. Project references crossing this boundary would need to be changed to PackageReferences, probably, which will be a pain for someone making a change in VS that needs a change in the core libs. It's not likely to bite us again until net6 is close to GA in a year from now, so I don't think it's worth going to the effort.

Nirmal4G replied:
I've removed the Preview 8 and RC 1 patches.

zivkan reviewed (commented):
The Windows jobs (unit tests and func tests) failed the "Install .NET 5.0 for build" step:

```text
. 'C:\A\1\71\s\scripts\utils\InstallCLIforBuild.ps1'   5.0.100-rc.1.20452.10
Downloading .NET SDK Install Script
Channel is : 5.0.100-rc.1.20452.10 latest
##[error]Invoke-RestMethod : BlobNotFoundThe specified blob does not exist.
RequestId:ce54500d-501e-007f-035a-9a36f9000000
Time:2020-10-04T14:29:41.3924590Z
At C:\A\1\71\s\scripts\utils\InstallCLIforBuild.ps1:47 char:20
+     $versionFile = Invoke-RestMethod -Method Get -Uri $httpGetUrl
+                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidOperation: (System.Net.HttpWebRequest:HttpWebRequest) [Invoke-RestMethod], WebExc 
   eption
    + FullyQualifiedErrorId : WebCmdletWebResponseException,Microsoft.PowerShell.Commands.InvokeRestMethodCommand

The version of SDK should be installed is : 
Probing folder : C:\A\1\_temp\dotnet\sdk\
C:\A\1\_temp\dotnet\dotnet-install.ps1 5.0.100-rc.1.20452.10 latest
##[error]Could not resolve version information.
At C:\A\1\_temp\dotnet\dotnet-install.ps1:323 char:9
+         throw "Could not resolve version information."
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (Could not resolve version information.:String) [], RuntimeException
    + FullyQualifiedErrorId : Could not resolve version information.
 

##[error]Process completed with exit code 1 and had 2 error(s) written to the error stream.
Finishing: Install .NET 5.0 for build
```


On Linux, the "Install .NET 5.0 for build" step was incorrectly reported as successful, but when I look at the logs I see:
```text
5.0.100-rc.1.20452.10
Channel is: 5.0.100-rc.1.20452.10    Version is: latest
Installing dotnet CLI into /home/vsts/work/_temp folder for building
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0

100 34939  100 34939    0     0   109k      0 --:--:-- --:--:-- --:--:--  109k
curl: (22) The requested URL returned error: 404 The specified blob does not exist.
Add /home/vsts/work/_temp/dotnet to PATH
/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/usr/share/rust/.cargo/bin:/home/runner/.config/composer/vendor/bin:/home/runner/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/home/vsts/work/_temp/dotnet
Deleting .NET Core temporary files
A compatible installed .NET Core SDK for global.json version [5.0.100-preview] from [/home/vsts/work/1/s/global.json] was not found
```

The `CliVersionForBuilding` property in config.props isn't used in `build.ps1`, so a successful local build won't test this. Locally, our scripts assume that the dev has installed an appropriate version themselves, and we use the system dotnet cli. On CI, we know the agents contain only GA versions of the dotnet sdk, so we have additional steps that only run on CI to get the preview SDK.

Nirmal4G commented:
@zivkan Can you restart the CI.

Nirmal4G commented:
@zivkan Can you restart the CI?

zivkan commented:
35 functional tests on Windows failed. All of them were Dotnet.Integration.Tests PackCommandTests. I'm sure they all failed for the same reason. One of them failed with this message:
```text
Pack failed with following log information :
Microsoft (R) Build Engine version 16.7.0-preview-20360-03+188921e2f for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

Determining projects to restore...
All projects are up-to-date for restore.
You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
C:\A\1\2\s.test\work\723a8b42\d2ee14af\sdk\5.0.100-preview.7.20366.6\Sdks\Microsoft.NET.Sdk\targets\Microsoft.PackageDependencyResolution.targets(241,5): error NETSDK1005: Assets file 'C:\A\1\2\s.test\work\192d45e0\291a9f42\ClassLibrary1\obj\project.assets.json' doesn't have a target for '.NETCoreApp,Version=v5.0'. Ensure that restore has run and that you have included 'net5.0' in the TargetFrameworks for your project. [C:\A\1\2\s.test\work\192d45e0\291a9f42\ClassLibrary1\ClassLibrary1.csproj]

Expected: True
Actual:   False
```

with this stack trace:
```
   at Dotnet.Integration.Test.MsbuildIntegrationTestFixture.PackProjectOrSolution(String workingDirectory, String file, String args, Boolean validateSuccess) in C:\A\1\2\s\test\NuGet.Core.FuncTests\Dotnet.Integration.Test\MsbuildIntegrationTestFixture.cs:line 273
   at Dotnet.Integration.Test.PackCommandTests.PackCommand_PackProject_OutputsBuildActionForContentFiles(String itemType, String buildAction, String expectedBuildAction) in C:\A\1\2\s\test\NuGet.Core.FuncTests\Dotnet.Integration.Test\PackCommandTests.cs:line 2655
```


Both Linux and Mac failed to even build the solution. The "Install .NET 5 for build" step on both have the same error message.
```text
master:5.0.100-preview.7.20366.6
Channel is: master:5.0.100-preview.7.20366.6    Version is: latest
Installing dotnet CLI into /home/vsts/work/_temp folder for building
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0

100 34939  100 34939    0     0  74655      0 --:--:-- --:--:-- --:--:-- 74655
curl: (22) The requested URL returned error: 404 The specified blob does not exist.
Add /home/vsts/work/_temp/dotnet to PATH
```

They both execute [`scripts/utils/InstallCLIforBuild.sh`](https://github.com/NuGet/NuGet.Client/blob/dev/scripts/utils/InstallCLIforBuild.sh), so that script probably needs similar changes to whatever you did in `InstallCLIforBuild.ps1`, which is used only on Windows.

Nirmal4G commented:
> They both execute [`scripts/utils/InstallCLIforBuild.sh`](https://github.com/NuGet/NuGet.Client/blob/dev/scripts/utils/InstallCLIforBuild.sh), so that script probably needs similar changes to whatever you did in `InstallCLIforBuild.ps1`, which is used only on Windows.

I thought It was using `.ps1` under the hood but it looks like a bash script unlike the arcade version which uses PowerShell all the way. I'll update them. Thanks for the update!

Nirmal4G commented:
> 35 functional tests on Windows failed. All of them were Dotnet.Integration.Tests PackCommandTests. I'm sure they all failed for the same reason....

Will investigate.

nkolev92 commented:
Hey @Nirmal4G 
I see this PR has been in draft for a while, as per https://github.com/NuGet/NuGet.Client/blob/dev/docs/workflow.md#draft-pull-requests, a PR should have a clear end game, and we're supposed to closed that draft PRs within 30 days. 

Closing per that. 

Feel free to re-open when this becomes relevant.

Nirmal4G commented:
@nkolev92 I can't re-open this PR. There is no option. Can you re-open this?

nkolev92 commented on .gitignore:
```diff
@@ -14,7 +14,6 @@ artifacts/
 **/.vscode/
 
 cli/
-cli1.0.4/
```
I wouldn't remove this just yet.
This was useful in earlier versions, and there's no harm in keeping it.

Some of us have to switch to older versions of code sometimes and it would mess with our builds.

Nirmal4G replied:
You have git for that, right?

zivkan replied:
The problem is that if I checkout an old branch and run `configure.ps1`, it will create the `cli1.0.4` folder, and be correctly ignored. Now, if I checkout dev, I still have `cli1.0.4`, but if `.gitignore` is no longer ignoring `cli1.0.4`, then it's going to want to track those files.

Nirmal4G replied:
Ahhh... I see now! 🤯

nkolev92 commented on build.sh:
```diff
@@ -15,19 +15,29 @@ echo "Installing dotnet CLI"
 DOTNET="$(pwd)/cli/dotnet"
 
-echo "$DOTNET msbuild build/config.props /v:m /nologo /t:GetCliBranchForTesting"
+echo "$DOTNET msbuild build/config.props /noAutoRsp /v:m /nologo /t:GetCliBranchForTesting"
```
do you have a rsp file in your workspace? 
I imagine the same thing would happen with directory.build.props/targets too.

Nirmal4G replied:
I thought the same thing. Even if there is a global rsp file, it would apply to the all the MSBuild calls except the above ones which may lead to inconsistent local builds.

nkolev92 commented on build/common.ps1:
```diff
@@ -157,95 +157,81 @@ Function Install-DotnetCLI {
-    $MSBuildExe = Get-MSBuildExe $vsMajorVersion
-    $CliBranchListForTesting = & $msbuildExe $NuGetClientRoot\build\config.props /v:m /nologo /t:GetCliBranchForTesting
-    $CliBranchList = $CliBranchListForTesting.Split(';');
+    $env:DOTNET_HOME = $CLIRoot
```
I'd only set the variable *after* the cli has been installed.

nkolev92 reviewed (commented):
Thanks for this PR. 

I appreciate that there's changes that should help others build locally, but there's a lot of changes that are just more difficult to review than the benefit said refactoring adds. 

We had a local version of the install script because downloading it was an issue on our CI servers, so unsure if we want to delete that yet - cc @zivkan

Nirmal4G commented:
I still have the next set of changes with local script removed. Let's see if it builds properly on the CI!

Nirmal4G commented:
I still have the next set of changes with local script removed. Let's see if these changes build properly on the CI!

Nirmal4G commented:
@zivkan Can you start the CI, please!?

Nirmal4G commented:
@nkolev92 I have removed the irrelevant changes (those'll come after this PR). This PR is now focused and to the point.

Nirmal4G commented:
@nkolev92 @zivkan 🔔

nkolev92 commented on build.sh:
```diff
@@ -15,19 +15,38 @@ echo "Installing dotnet CLI"
+# Get CLI Branches for testing
+echo "dotnet msbuild build/config.props /v:m /nologo /t:GetCliBranchForTesting"
+CMD_OUT_LINES=(`$DOTNET msbuild build/config.props /v:m /nologo /t:GetCliBranchForTesting`)
+DOTNET_BRANCHES=${CMD_OUT_LINES[-1]}
```
what's the purpose of 45 and 46?
Why not assign to $DOTNET_BRANCH directly?

Nirmal4G replied:
This is the fix

Nirmal4G replied:
The next patch was not yet pushed at that time. It adds the version logic from `runFuncTests.sh`. I did this way to make diff between commits easy to understand visually.

nkolev92 commented on build/common.ps1:
```diff
@@ -160,7 +160,7 @@ Function Install-DotnetCLI {
     $MSBuildExe = Get-MSBuildExe $vsMajorVersion
     $CliBranchListForTesting = & $msbuildExe $NuGetClientRoot\build\config.props /v:m /nologo /t:GetCliBranchForTesting
-    $CliBranchList = $CliBranchListForTesting.Split(';');
+    $CliBranchList = $CliBranchListForTesting[-1] -split ';'
```
Is this the -1 a typo?

Nirmal4G replied:
No, that was the fix. But seems to work only when `rsp` file is present.

nkolev92 replied:
Not sure I follow. If I do the following: 
```powershell

$CliBranchListForTesting = & MSBuild.exe build\config.props /v:m /nologo /t:GetCliBranchForTesting
$CliBranchList = $CliBranchListForTesting.Split(';');
$CliBranchList2 = $CliBranchListForTesting[-1] -split ';'

echo "Original: '$CliBranchList'"
echo "New: '$CliBranchList2'"
```

The output is
```
Original: '  master 5.0.100'
New: '0'
```

Nirmal4G replied:
fixed it, again!

Nirmal4G replied:
2nd attempt at fixing this. please check again.

nkolev92 reviewed (commented):
A couple of little comments.

Not sure if this change does much at this point. This looks like couple of stylistic changes at this point.

zivkan commented:
The linked issue that this PR claims to fix is about `rsp` files, but I don't see anything in the change set that appears to resolve that. I'm not sure how to validate the "correctness" of this PR.

Also, did you test `build.sh` locally?  Our CI doesn't run it (it does run `runFuncTests.sh`). We've had at least two customers tell us that `build.sh` doesn't work, so I'm not sure if this PR fixes that.

Nirmal4G commented:
@nkolev92 @zivkan

`build.sh` downloads dotnet cli v1.0 for bootstrapping and runs msbuild to get the version specified, right?

But the issue is 1.0 is out of support and doesn't get downloaded, so, we have to update to least min version to 2.1 but that reads global json leading to cli version mismatch.

Either we have to update the link to get v1 cli or we have to update cli acquiring logic. if we're moving to Arcade then the former one is the simplest solution.

Nirmal4G commented:
@nkolev92 @zivkan

`build.sh` downloads `dotnet` cli v1.0 for bootstrapping and runs `msbuild` to get the version specified, right?
But the issue is 1.0 is out of support and doesn't get downloaded, so, we have to update to least min version to 2.1 but that reads `global.json` leading to cli version mismatch.
So, I have updated the min bootstrapping version to 5.0 to match `global.json` and install the 2.1 runtime later in the setup.

I have also tried using `global.json` to install the bootstrapping version directly. Since it has `-preview` in it, the install fails with this message...

```log
dotnet_install: Error: Could not find `.NET Core SDK` with version = 5.0.100-preview
dotnet_install: Error: Refer to: https://aka.ms/dotnet-os-lifecycle for information on .NET Core support
```

Either we have to update the link to get v1 cli or we have to update cli acquiring logic. if we're moving to Arcade then the former one is the simplest solution.

Nirmal4G commented:
Made some changes. Please check to see if it works locally... Both PowerShell and bash scripts.

Nirmal4G commented:
Made some changes. Please check to see if it works locally... Both PowerShell and bash scripts. I stopped at restoring since the changes are only in setup.

Also, my machine is not powerful enough to run full build. That's why I keep asking for CI. Sorry about that.

zivkan commented:
Hi @Nirmal4G I'm really sorry about the delays reviewing this PR. We've had some very busy weeks.

Given that this pull request does not impact customers using NuGet as a product, it's unfortunately lower priority for us. If you'd like to take up a customer impacting issue, we have several [with the "help wanted" label](https://github.com/NuGet/Home/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22).

But given that our builds are not public, so any failures require us to check the log, copy & paste the error here for you to see, plus the time to review the changes, it's time that we cannot commit to currently.

Nirmal4G commented:
@zivkan

> I'm really sorry about the delays...

Don't be. I'm not worried about it.

> Given that this pull request does not impact customers using NuGet as a product...

True, but really impacts me as a contributor.

I have a Windows laptop (not powerful enough) but I do have a Linux workstation. I also have a response file (_to have a local friendly build_) in the top of all repos I work with. For this repo, I have to work on top of this patch and rebase or cherry-pick for every change into PRs (_like #3270_) because of these issues.

It's really a pain in the...😩

> If you'd like to take up a customer impacting issue, we have several with the "help wanted" label.

It seems that this one is associated with an issue marked with a "help wanted" label. I didn't even know until you showed me those. 😏

> ...But given that our builds are not public...

Please make one public pipeline that builds and tests PRs (_similar to what we do locally_) for every push at least for contributors who need the label `Approved for CI`.

zivkan commented:
> For this repo, I have to work on top of this patch and rebase or cherry-pick for every change

I wasn't aware. The code changes looked stylistic, rather than functional. Especially the `runFuncTests.sh` changes, which run in our CI builds, and currently work for us without error. So, it's not clear to me why you're modifying that file, and if you're encountering problems running it, why it works on our CI, but not your local machine. Since several of the changes to `build.sh` look the same as the changes made to `runFuncTests.sh`, I assumed they similarly would not fix build errors.

If you could stop changing the scope of this PR (first it depended on another PR being merged, then it's about solving a problem with .rsp files, now it's something else), then we won't have a moving target, making it easier to review and merge.

Can you please review your PR title and description. The title "fix issues" isn't very descriptive. The desciption and list of issues this PR claims to fix doesn't seem to match the contents of the PR either. Or, I don't understand how the changes prevent the .rsp file issue.

Also, please remove the changes you made regarding workarounds for issue 8936. If you feel strongly about it, create it in a separate PR, because if we merge it and we start having CI failures again, I'm going to revert the commit and it's going to undo all the other changes you made in this PR. In general prefer keeping one PR specific to one issue, so that it can be cherry-picked and reverted with minimal side-effects.

>Please make one public pipeline that builds and tests PRs

I'd love to, but it takes effort, which I don't have time for right now without stopping work on other deliverables I have. We're making progress, it's just a question of time and prioritization.

Nirmal4G commented:
<details>
<summary><em>Reply</em></summary>
<br/>

> If you could stop changing the scope of this PR...

Never did, from the first push. The PR was first opened for fixing only `rsp` issue. The initial patch included some **refactoring** basing on top of Heng Liu's patch but Nikolche Kolev suggested to remove it as it was dropped. Then, I found that bash scripts can't run properly on WSL because of the line endings (_1st commit_). Also, `build.sh` wasn't working properly when I tested the repo in WSL. So, I fixed those issues along with the removal of the local `dotnet-install.sh` script doesn't work properly in WSL either.

Just as the branch name and the issue details, the focus has always been about making the scripts work locally with `rsp` file and in WSL.

> Also, please remove the changes you made regarding workarounds for issue 8936. If you feel strongly about it, create it in a separate PR...

Only the last commit have that change. So, yes, I'll separate that into a new PR.

> In general, prefer keeping one PR specific to one issue...

This PR is mostly about fixing `rsp` issue (_3rd commit_). But the `build.sh` fix (_4th commit_) is just changing the bootstrapping version of cli from v1.0 to v2.2 and matching the cli install logic from `runFuncTests.sh`!

Also, `build.sh` script doesn't run on CI. That's why, I included here.

</details>

In summary, I could separate this into 2 or 3 separate PRs. One for the `git checkout` and `build.sh` fix, one for the `rsp` fix and one for removing the local `dotnet-install.sh` script.

_How do you want to proceed?_

zivkan commented:
My crunch time is hopefully coming to an end, so I took some time to slowly read the linked issues. Specifically this line:

> when the response file is detected, MSBuild spews out a header text for the response file

I created a `Directory.Build.rsp` file on my machine, but it's not changing the output, unless I put `-v:d` or something similar. But since our scripts explicitly use `/v:m` on the command line, it overrides the `rsp` verbosity, hence the output isn't modified. Is it possible a different rsp file will change the end of out output, rather than the beginning, like in your case?

Honestly, my gut reaction is to create empty `Directory.Build.rsp` file in our repo root, so this filename will never be read from outside the repo.

However, I'm starting to understand the changes, specifically splitting on the newline character and reading the `[-1]` element.

I still believe it's best to move the `dotnet-install.sh` changes to a different PR, because I'm not sure our CI agents won't just have whatever error caused us to create issue 8936 and the workaround (commit the script into our repo) in the first place. If it's still a problem, then it would block your other changes. If it succeeds now with the workaround removed, but then has the problem again in a month, then we'll revert the change and lose any other changes in the same commit.

Nirmal4G commented:
@zivkan I've removed the `dotnet-install.sh` changes from this PR.

Nirmal4G commented on build/config.props:
```diff
@@ -34,7 +34,7 @@
     Note that multiple SDKs can be downloaded by using `;` as a separator.
     -->
-    <CliBranchForTesting Condition="'$(CliBranchForTesting)' == ''">release/5.0.2xx 5.0.200-servicing.21120.4</CliBranchForTesting>
+    <CliBranchForTesting Condition="'$(CliBranchForTesting)' == ''">release/5.0.2xx:5.0.200-servicing.21120.4</CliBranchForTesting>
```
Changed the space ` ` separator into colon `:` separator. Only way to make the `IFS` work properly with for loop.

zivkan commented on build.sh:
```diff
@@ -15,26 +15,61 @@ echo "Installing dotnet CLI"
+
+	# Display current version
+	$DOTNET --version
+	dotnet --info
```
This should be `$DOTNET --info`, so we see the information from the .NET SDK installed in the `cli/` directory. With this change, it's only showing system installed SDKs and runtimes, so the logs don't help us confirm that the .NET SDK was correctly installed into the `cli/` folder.

Personally, I think we should also set `DOTNET_MULTILEVEL_LOOKUP=0`, and maybe `DOTNET_HOME`, or whatever other environment variables are important to the dotnet cli, so it doesn't also list the system installed SDK and runtimes. But we don't do that anywhere else.

zivkan commented:
@Nirmal4G I rebased your PR so it includes some test fixes. Other people's PR builds are green now, but your branch has two errors.

On Mac, the log has this output:

```text
  Discovering test assemblies for NuGet.Protocol.FuncTest
  NuGet.CommandLine.XPlat -> /Users/runner/work/1/s/artifacts/NuGet.CommandLine.XPlat/16.0/bin/Release/netcoreapp2.1/NuGet.CommandLine.XPlat.dll
  NuGet.CommandLine.Xplat.Tests -> /Users/runner/work/1/s/test/NuGet.Core.Tests/NuGet.CommandLine.Xplat.Tests/bin/Release/netcoreapp2.1/NuGet.CommandLine.Xplat.Tests.dll
  Microsoft.Build.NuGetSdkResolver -> /Users/runner/work/1/s/artifacts/Microsoft.Build.NuGetSdkResolver/16.0/bin/Release/netstandard2.0/Microsoft.Build.NuGetSdkResolver.dll
  NuGet.XPlat.FuncTest -> /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/bin/Release/netcoreapp2.1/NuGet.XPlat.FuncTest.dll
  Discovering test assemblies for NuGet.XPlat.FuncTest
MSBUILD : warning MSB1008: Only one project can be specified. [/Users/runner/work/1/s/build/build.proj]
  Switch: /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/bin/Release/netcoreapp2.1/NuGet.Commands.FuncTest.dll
  
  For switch syntax, type "MSBuild -help"
/Users/runner/work/1/s/build/build.proj(439,5): warning MSB3073: The command "/Users/runner/work/1/s/cli/dotnet test /Users/runner/work/1/s/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/bin/Release/netcoreapp5.0/Dotnet.Integration.Test.dll /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/bin/Release/netcoreapp2.1/NuGet.Commands.FuncTest.dll /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.Commands.FuncTest/bin/Release/netcoreapp5.0/NuGet.Commands.FuncTest.dll /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/bin/Release/netcoreapp2.1/NuGet.Packaging.FuncTest.dll /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.Packaging.FuncTest/bin/Release/netcoreapp5.0/NuGet.Packaging.FuncTest.dll /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/bin/Release/netcoreapp2.1/NuGet.Protocol.FuncTest.dll /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.Protocol.FuncTest/bin/Release/netcoreapp5.0/NuGet.Protocol.FuncTest.dll /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/bin/Release/netcoreapp2.1/NuGet.XPlat.FuncTest.dll /Users/runner/work/1/s/test/NuGet.Core.FuncTests/NuGet.XPlat.FuncTest/bin/Release/netcoreapp5.0/NuGet.XPlat.FuncTest.dll --test-adapter-path:/Users/runner/work/1/s/packages/xunitxml.testlogger/2.0.0/build/_common --logger:"xunit;LogFileName=CoreFuncTests-vsts.xml" --logger:"console;verbosity=detailed" --settings:/Users/runner/work/1/s/build/xunit.runsettings" exited with code 1.
/Users/runner/work/1/s/build/build.proj(439,5): warning MSB4181: The "Exec" task returned false but did not log an error.
/Users/runner/work/1/s/build/build.proj(446,5): error : NETCore CoreFuncTests tests failed! Results: /Users/runner/work/1/s/build/TestResults/CoreFuncTests-vsts.xml
CoreFuncTests failed!!
Checking if result file exists at /Users/runner/work/1/s/scripts/funcTests/../../build/TestResults/TestResults.xml
/Users/runner/work/1/s/scripts/funcTests/../../build/TestResults/TestResults.xml not found.
```

I'm not 100% sure it's related to your changes, but I've also never before seen xunit discover the tests, but then not actually execute any of them.

On Linux, tests run, but 17 `Dotnet.Intregration.Tests` fail with this error:

```text
System.AggregateException : One or more errors occurred. (Sequence contains more than one element) (The following constructor parameters did not have matching fixture data: MsbuildIntegrationTestFixture fixture)
---- System.InvalidOperationException : Sequence contains more than one element
---- The following constructor parameters did not have matching fixture data: MsbuildIntegrationTestFixture fixture
```

```text
----- Inner Stack Trace #1 (System.InvalidOperationException) -----
   at System.Linq.ThrowHelper.ThrowMoreThanOneElementException()
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at Dotnet.Integration.Test.MsbuildIntegrationTestFixture..ctor() in /home/vsts/work/1/s/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs:line 42
----- Inner Stack Trace #2 (Xunit.Sdk.TestClassException) -----
```

This class, `MsbuildIntegrationTestFixture` is responsible for copying the .NET SDK into a test folder, and replacing all the NuGet DLLs with the NuGet DLLs from the build. Therefore, we can test the .NET SDK with the current build of NuGet, not the version of NuGet that shipped in the .NET SDK.

As a sanity check, it makes sure that the test SDK only has a single folder in the SDK subdirectory, but for some reason there's more than one. Frankly, I have no guesses why this is happening, but this happened for at least two builds of this branch, and other people's branches are passing, so it appears related to the changes in this PR.

Nirmal4G committed:
Use global and local dotnet CLI properly

Since the test frameworks use copied and patched cli, mismatch can cause tests fail.

One of the failures caused when sometimes "NuGetFallbackFolder" becomes the first directory
to resolve "dotnet.dll" against. It is now fixed by excluding the fallback folder in the resolving logic.

zivkan commented on scripts/funcTests/runFuncTests.sh:
```diff
@@ -135,7 +158,7 @@ echo "Core tests finished at `date -u +"%Y-%m-%dT%H:%M:%S"`"
 # Func tests
 echo "dotnet msbuild build/build.proj /t:CoreFuncTests /p:VisualStudioVersion=16.0 /p:Configuration=Release /p:BuildNumber=1 /p:ReleaseLabel=beta"
-dotnet msbuild build/build.proj /t:CoreFuncTests /p:VisualStudioVersion=16.0 /p:Configuration=Release /p:BuildNumber=1 /p:ReleaseLabel=beta
+$DOTNET msbuild build/build.proj /t:CoreFuncTests /p:VisualStudioVersion=16.0 /p:Configuration=Release /p:BuildNumber=1 /p:ReleaseLabel=beta
```
The .NET SDK installed into the `cli/` folder is used by `Dotnet.Integration.Tests`. I don't think we want to use it to build the product though.

Nirmal4G replied:
So, we use System installed .NET SDK to build the product and local one to test the product, right?

zivkan replied:
We don't use the local SDK to run `dotnet test`, we still use the system SDK for that. Rather, [Dotnet.Integration.Test's test fixture](https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs) copies the SDK from the `cli/` folder to a `.test\<random>` folder and then replaces all the NuGet assemblies, then uses it itself.  Ideally, `Dotnet.Integration.Test` would download the SDK itself, so when `Dotnet.Integration.Test` is not run, there's no need to download the SDK into the `cli/` folder. But we're not there yet. In case you feel interested in implementing this, please do not do it in this PR.

I think the script to download the .NET SDK into the `cli/` folder before running other commands was from the pre-.NET Core 1.0 days, when CI agents did not have the .NET SDK pre-installed.  Now that they do, I believe these build/run tests scripts should use the system `dotnet` for everything, except for the `cli/dotnet --info` that helps us validate the build is working as expected.

But leaving each command to use whichever dotnet it's already using is a safe bet, as it minimizes unintended consequences. Let's focus our efforts on improvements that fix reported problems.

Nirmal4G replied:
> We don't use the local SDK to run `dotnet test`

That, I understand. Some calls in the script had `$DOTNET` instead. That's why I was confused. Thought, that might be the problem. I can reproduce the issue in my WSL Ubuntu setup. I don't have a Mac. So, I don't know how to test in a Mac environment. But since both are Unix Like environments, fixing in Linux should do the same in Mac, I think!

> ...please do not do it in this PR...

Don't worry! This patch is complicated already!! 😂 But I'm interested in that change. I'll look into it after this.

Nirmal4G commented:
@zivkan My Build on WSL succeeds. Please check with CI now.

zivkan commented:
Now only tests on mac are failing.

Here's the full build logs: 
[mac-build-log.zip](https://github.com/NuGet/NuGet.Client/files/6225291/mac-build-log.zip)


Looks like I was wrong about running tests using the system dotnet. At `2021-03-29T23:13:24.4407250Z` we see that it's running `cli/dotnet test ...`, and the next 3 lines are saying that command failed, but there's no output from xunit, hence it never actually ran.

At `2021-03-29T23:10:22.5364700Z`, we see that the .NET SDK installed in the `cli/` directory only has 2.2, there's no 5.0 SDK.  That's my guess as to why the unit tests and func tests failed to run. But there's no hint as to why it failed to install the SDK defined in the `build/config.props` file.

I just remembered that months/years ago, I read something saying that Apple had a very old veresion of Bash on OSX, and they also changed the default shell to zsh. I don't know if our CI builds on Mac are using zsh, or bash, and if it's using bash, what version. Although you might be able to figure it out from the build logs. If your changes to `runFuncTests.sh` don't work on zsh, or require a new version of Bash, perhaps that's why it's failing to install the .NET 5 SDK, although it's disappointing that it's silent.

Like you, I don't have access to a Mac, at least, not until we're allowed back in the office. So, we're at the mercy of someone with a Mac testing for us. Alternatively, you could do more research into what shell Azure DevOps runs scripts on, then run the same shell in your linux/wsl install, and see if the script fails like on CI, allowing you to test locally.

Another option is to undo all your changes and just create an empty response file in the repo root. I don't know how a response file helps you manage your builds/repos, but creating an empty response file in NuGet.Client's root should prevent msbuild finding the response file outside the repo root, and be a lower complexity solution to the problem.

Nirmal4G commented on scripts/funcTests/runFuncTests.sh:
```diff
@@ -36,18 +38,36 @@ chmod +x scripts/funcTests/dotnet-install.sh
+IFS=$'\n'
+CMD_OUT_LINES=(`dotnet msbuild build/config.props /v:m /nologo /t:GetCliBranchForTesting`)
+# Take the line just before the last empty line and remove all the spaces
+DOTNET_BRANCHES=${CMD_OUT_LINES[-1]//[[:space:]]}
```
Seems this is causing mac build to fail. Here's what I found from the log…

<details>
<summary>
You can see that this line fails with a <code>bad array subscript</code> error.
</summary>
<br/>

```log
•••
dotnet msbuild build/config.props /v:m /nologo /t:GetCliBranchForTesting

/Users/runner/work/1/s/scripts/funcTests/runFuncTests.sh: line 62: CMD_OUT_LINES: bad array subscript

.NET Core SDK (reflecting any global.json):
 Version:   2.2.402
•••
.NET Core SDKs installed:
  2.2.402 [/Users/runner/work/1/s/cli/sdk]
•••
```
</details>

You are right with having old version of Bash on macOS. Bash on macOS is v3.2 but only v4.3+ supports negative indexing. That's why it errors out. The following change should fix it…

```suggestion
CMD_LAST_LINE=${CMD_OUT_LINES[@]:(-1)}
DOTNET_BRANCHES=${CMD_LAST_LINE//[[:space:]]}
```

<details>
<summary>More info on this line</summary>
<br/>

This logic takes the last line from the `GetCliBranchForTesting` target and strips out any whitespace across the line.

MSBuild already outputs with an indent, and you could also have whitespace within the string. So, you’ll have an output string like these…

1. `    release/5.0.2xx : 5.0.201`
2. `    master ; release/5.0.2xx : 5.0.201`

Which are perfectly valid but POSIX shell doesn't srtip leading/trailing spaces when assigned to a variable leading to a bunch of issues.

One of the reasons I used colon (`:`) as a separator to avoid splitting by spaces which could result in empty entries in the array depending on the behavior of split on each platform.

It's not like in `PowerShell` where you can trim leading/trailing whitespace using `string.Trim()` method.

In summary, this fix is recommended.
</details>

Nirmal4G reviewed (commented):
> Another option is to undo all your changes and just create an empty response file in the repo root.

This doesn't work. It doesn't matter where the response file is but as long as there's one that MSBuild can find and load, it displays that message and fails to get the version from the `GetCliBranchForTesting` target.

> There should be a better way to get values from MSBuild targets.

Nirmal4G committed:
Fix macOS build failure

Bash on macOS is v3.2 but only v4.3+ supports negative indexing.

This caused the script skipping the .NET SDK install for to be used with .NET CLI integration tests,
which in-turn caused the build to fail. This is a verified workaround that works on Bash versions < v4.3

Nirmal4G commented:
@zivkan Fixed the MacOS build. Please check with the CI.

Nirmal4G commented:
@zivkan Fixed the MacOS build. Please check with the CI.

zivkan commented:
> This doesn't work. It doesn't matter where the response file is but as long as there's one that MSBuild can find and load, it displays that message and fails to get the version from the GetCliBranchForTesting target.

What output?  When I tested myself I didn't see any extra output, and looking at MSBuild's source code, I only see extra output at diagnostic verbosity, not minimal verbosity: https://github.com/dotnet/msbuild/blob/1629921c2b537d4452ff238981bd18519b8f5230/src/MSBuild/XMake.cs#L2316-L2319

Nirmal4G commented:
I looked at the source too. I don't know why or how but this is how, it outputs for me…!

```cmd
nuget-client > msbuild build\config.props -nologo -v:q -t:GetCliBranchForTesting

Some command line switches were read from the auto-response file "MSBuild.rsp". To disable this file, use the "-noAutoResponse" switch.
D:\Apps\Tools\MSBuild\Current\Bin\msbuild.exe -nologo -bl -t:GetCliBranchForTesting -v:q build\config.props
```
```cmd
nuget-client > msbuild build\config.props -nologo -v:m -t:GetCliBranchForTesting

Some command line switches were read from the auto-response file "MSBuild.rsp". To disable this file, use the "-noAutoResponse" switch.
D:\Apps\Tools\MSBuild\Current\Bin\msbuild.exe -nologo -bl -t:GetCliBranchForTesting -v:m build\config.props
  release/5.0.2xx 5.0.200-servicing.21120.4
```
```cmd
nuget-client > msbuild build\config.props -nologo -v:n -t:GetCliBranchForTesting

Some command line switches were read from the auto-response file "MSBuild.rsp". To disable this file, use the "-noAutoResponse" switch.
D:\Apps\Tools\MSBuild\Current\Bin\msbuild.exe -nologo -bl -t:GetCliBranchForTesting -v:n build\config.props
Build started 31-03-2021 12:06:45 AM.
Project "E:\Projects\Microsoft.NET\nuget-client\build\config.props" on node 1 (GetCliBranchForTesting target(s)).
GetCliBranchForTesting:
  release/5.0.2xx 5.0.200-servicing.21120.4
Done Building Project "E:\Projects\Microsoft.NET\nuget-client\build\config.props" (GetCliBranchForTesting target(s)).

•••

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.09
```

Nirmal4G commented:
Finally, All ✔️. 🥳
